proposals/0111-linear-types.rst

```diff
@@ -907,6 +909,159 @@ or ``case`` in the surface syntax), we want to infer the multiplicity
 annotation. The process for this is not yet defined (see `Unresolved
 questions`_ below for a more precise description of this issue).
 
+Let and where bindings
+----------------------
+
+*This section is written with let bindings, but all of the same applies
+to where bindings.*
+
+Note on terminology: following the Haskell 2010 report, a function
+binding is a binding of the form ``f arg1…argn =`` with at least one
+arguments. The other binding is called pattern bindings. In particular
+``let x =`` is a pattern binding (as opposed to how it is in GHC's
+implementation, where ``let x =`` would be a ``FunBind``). This
+proposal will be using the terminology “non-variable pattern”
+for patterns which aren't a single variable.
+
+Let bindings can optionally be annotated with a multiplicity
```

Comment:
I like conservatism in this way in general, but I tend to think this is an easy place to accept a little more. I don't feel strongly.

Reply:
I don't feel very strongly either. I've added an alternative to that effect.

Because rendering is broken, let me copy in full what I wrote:

> Restrictions of multiplicity-annotated let bindings
> ---------------------------------------------------
> 
> The proposal specifies that a multiplicity annotated non-variable let binding ``let %p pat``
> must be such that ``pat = !pat'`` even if ``p = 'Many``. It is easy to
> lift this restriction on two dimension:
> 
> - We can say, instead, that patterns not of the form ``!pat'`` emit a
>   ``p ~ 'Many`` constraint instead. They already do (for the sake of
>   inference), so this is strictly less code.
> - We can generalise to more strict patterns. For instance, we don't
>   need to require a ``!`` if ``-XStrict`` is on, we can have patterns
>   of the form ``(!pat')`` (with additional parentheses). This is a few
>   lines of codes, inference actually already does this in my
>   implementation, so it's already paid for (though it does annoyingly
>   mostly duplicate another definition of strict pattern which I
>   couldn't find a way to factor as a single function, I don't like
>   this).
> 
> The reason that motivated the stronger restriction is to improve error
> messages, because we can then error out with “multiplicity-annotated
> let-bound patterns must be of the form !pat”, instead of the more
> mysterious “Couldn't unify 'Many with 'One”
> (see [#23586](https://gitlab.haskell.org/ghc/ghc/-/issues/23586)).
> But maybe the additional restrictions are more surprising than the
> error messages are helpful.